src/track/serato/beatgrid.h

```diff
@@ -0,0 +1,185 @@
+#pragma once
+
+#include <QByteArray>
+#include <QColor>
+#include <QDataStream>
+#include <QList>
+#include <memory>
+
+#include "track/cueinfo.h"
+#include "track/taglib/trackmetadata_file.h"
+#include "util/assert.h"
+#include "util/types.h"
+
+namespace mixxx {
+
+class SeratoBeatGridNonTerminalMarker;
+class SeratoBeatGridTerminalMarker;
+
+typedef std::shared_ptr<SeratoBeatGridNonTerminalMarker> SeratoBeatGridNonTerminalMarkerPointer;
+typedef std::shared_ptr<SeratoBeatGridTerminalMarker> SeratoBeatGridTerminalMarkerPointer;
+
+class SeratoBeatGridNonTerminalMarker {
+  public:
+    SeratoBeatGridNonTerminalMarker(float positionSecs, quint32 beatsTillNextMarker)
+            : m_positionSecs(positionSecs),
+              m_beatsTillNextMarker(beatsTillNextMarker) {
+    }
+    ~SeratoBeatGridNonTerminalMarker() = default;
+
+    QByteArray dumpID3() const;
+    static SeratoBeatGridNonTerminalMarkerPointer parseID3(const QByteArray&);
+
+    float positionSecs() const {
+        return m_positionSecs;
+    }
+
+    float beatsTillNextMarker() const {
+        return m_beatsTillNextMarker;
+    }
+
+  private:
+    float m_positionSecs;
+    quint32 m_beatsTillNextMarker;
+};
+
+inline QDebug operator<<(QDebug dbg, const SeratoBeatGridNonTerminalMarker& arg) {
+    return dbg << "SeratoBeatGridNonTerminalMarker"
+               << "PositionMillis =" << arg.positionSecs()
+               << "BeatTillNextMarker = " << arg.beatsTillNextMarker();
+}
+
+class SeratoBeatGridTerminalMarker {
+  public:
+    SeratoBeatGridTerminalMarker(float positionSecs, float bpm)
+            : m_positionSecs(positionSecs),
+              m_bpm(bpm) {
+    }
+    ~SeratoBeatGridTerminalMarker() = default;
+
+    QByteArray dumpID3() const;
+    static mixxx::SeratoBeatGridTerminalMarkerPointer parseID3(const QByteArray&);
+
+    float positionSecs() const {
+        return m_positionSecs;
+    }
+
+    float bpm() const {
+        return m_bpm;
+    }
+
+  private:
+    float m_positionSecs;
+    float m_bpm;
+};
+
+inline QDebug operator<<(QDebug dbg, const SeratoBeatGridTerminalMarker& arg) {
```

Comment:
I prefer to put these non-performance-critical utility functions into the .cpp file instead of cluttering the header file. Also when inlined the string literals might be duplicated on every call site (didn't check if the compiler is smart enough to recognize the redundancy).